pkg/sql/opt/xform/coster.go

```diff
@@ -254,6 +261,52 @@ func (c *coster) computeSortCost(sort *memo.SortExpr, required *physical.Require
 	return cost
 }
 
+func (c *coster) computeSegmentedSortCost(
+	sort *memo.SortExpr, required *physical.Required,
+) memo.Cost {
+	// We calculate the cost of a segmented sort. We assume each segment
+	// is of the same size of (rowCount / numOrderedSegments). We also calculate the
+	// per-row cost. The cost of the sort is:
+	//
+	// perRowCost * (rowCount + rowCount * log2(segmentSize))
+	//
+	// Or equivalently:
+	//
+	// perRowCost * (rowCount + ((segmentSize * log2(segmentSize)) * numOrderedSegments))
+	var numOrderedSegments float64
+	stats := sort.Relational().Stats
+	rowCount := stats.RowCount
+	perRowCost := c.rowSortCost(len(required.Ordering.Columns), len(sort.InputOrdering.Columns))
+	orderedCols := sort.InputOrdering.ColSet()
+	orderedStats, ok := stats.ColStats.Lookup(orderedCols)
+	if !ok {
+		orderedStats, ok = c.mem.RequestColStat(sort, orderedCols)
+		if !ok {
+			// I don't think we can ever get here.
+			//
+			// We can't get the ColStat because we don't have the stats
+			// builder (it has been cleared when SetRoot was called).
+			// If that happened though, the ColStat we're looking for
+			// must've been computed and stored when the coster was invoked
+			// the first time. And so the initial lookup should have succeeded.
+
+			// Also, side node: When is the coster blown away? Is this
+			// even reachable? Do we ever have a coster but not the stats builder?
+
+			// TODO(ridwanmsharif): Confirm this?
+		}
+	}
+
+	numOrderedSegments = orderedStats.DistinctCount
+	segmentSize := rowCount / numOrderedSegments
+	var cost memo.Cost
+	if segmentSize > 1 {
+		cost = memo.Cost(segmentSize) * (memo.Cost(math.Log2(segmentSize)) * memo.Cost(numOrderedSegments))
```

Comment:
Assuming the formula is correct, how about reducing floating point inaccuracies this way?
```go
if rowCount > numOrderedSegments {
  cost = memo.Cost(rowCount * (math.Log2(rowCount) - math.Log2(orderedStats.DistinctCount)))
}
```

(I'm not familiar with the project, would `orderedStats.DistinctCount` ever be non-positive?)